pkg/controllers/cronfederatedhpa/cronfederatedhpa_job.go

```diff
@@ -0,0 +1,280 @@
+
+	namespaceName types.NamespacedName
+	rule          autoscalingv1alpha1.CronFederatedHPARule
+}
+
+func NewCronFederatedHPAJob(client client.Client, eventRecorder record.EventRecorder,
+	namespaceName types.NamespacedName, rule autoscalingv1alpha1.CronFederatedHPARule) *CronFederatedHPAJob {
+	return &CronFederatedHPAJob{
+		client:        client,
+		eventRecorder: eventRecorder,
+		namespaceName: namespaceName,
+		rule:          rule,
+	}
+}
+
+func (c *CronFederatedHPAJob) Run() {
+	klog.V(4).Infof("Start to handle CronFederatedHPA %s", c.namespaceName)
+	defer klog.V(4).Infof("End to handle CronFederatedHPA %s", c.namespaceName)
+
+	var err error
+	cronFHPA := &autoscalingv1alpha1.CronFederatedHPA{}
+	err = c.client.Get(context.TODO(), c.namespaceName, cronFHPA)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			klog.Infof("CronFederatedHPA %s not found", c.namespaceName)
+		} else {
+			klog.Errorf("Get CronFederatedHPA failed: %v", err)
+		}
+		return
+	}
+
+	// It shouldn't be true, because we will not add the rule to schedule queue
+	if helper.IsCronFederatedHPARuleSuspend(c.rule) {
+		klog.V(4).Infof("CronFederatedHPA %s Rule %s is suspended", c.namespaceName, c.rule.Name)
+		return
+	}
+
+	var scaleErr error
+	defer func() {
+		if scaleErr != nil {
+			c.eventRecorder.Event(cronFHPA, corev1.EventTypeWarning, "ScaleFailed", scaleErr.Error())
+			err = c.addFailedExecutionHistory(cronFHPA, scaleErr.Error())
+		} else {
+			err = c.addSuccessExecutionHistory(cronFHPA, c.rule.TargetReplicas, c.rule.TargetMinReplicas, c.rule.TargetMaxReplicas)
+		}
+		if err != nil {
+			c.eventRecorder.Event(cronFHPA, corev1.EventTypeWarning, "ScaleFailed", err.Error())
+		}
+	}()
+
+	if cronFHPA.Spec.ScaleTargetRef.APIVersion == autoscalingv1alpha1.GroupVersion.String() {
+		if cronFHPA.Spec.ScaleTargetRef.Kind != autoscalingv1alpha1.FederatedHPAKind {
+			scaleErr = fmt.Errorf("do not support scale target %s/%s", cronFHPA.Spec.ScaleTargetRef.APIVersion, cronFHPA.Spec.ScaleTargetRef.Kind)
+			return
+		}
+
+		scaleErr = retry.RetryOnConflict(retry.DefaultRetry, func() (err error) {
+			err = c.ScaleFHPA(cronFHPA)
+			return err
+		})
+		return
+	}
+
+	// scale workload directly
+	scaleErr = retry.RetryOnConflict(retry.DefaultRetry, func() (err error) {
+		err = c.ScaleWorkloads(cronFHPA)
+		return err
+	})
+}
+
+func (c *CronFederatedHPAJob) ScaleFHPA(cronFHPA *autoscalingv1alpha1.CronFederatedHPA) error {
+	fhpaName := types.NamespacedName{
+		Namespace: cronFHPA.Namespace,
+		Name:      cronFHPA.Spec.ScaleTargetRef.Name,
+	}
+
+	fhpa := &autoscalingv1alpha1.FederatedHPA{}
+	err := c.client.Get(context.TODO(), fhpaName, fhpa)
+	if err != nil {
+		return err
+	}
+
+	update := false
+	if c.rule.TargetMaxReplicas != nil && fhpa.Spec.MaxReplicas != *c.rule.TargetMaxReplicas {
+		fhpa.Spec.MaxReplicas = *c.rule.TargetMaxReplicas
+		update = true
+	}
+	if c.rule.TargetMinReplicas != nil && *fhpa.Spec.MinReplicas != *c.rule.TargetMinReplicas {
```

Comment:
Just a question, if the user has an `fhpa` and its `MinReplicas` is equal to `10`, can the user use `cronfhpa` to set the `MinReplicas` of `fhpa` to `nil`?

Reply:
No, I think.